### ISSUE TITLE
docker container is privileged tries to assume more capabilities than available

### ISSUE THREAD

**Issue by smira** [moby/moby#42906]
tl;dr is that `docker run --privileged` tries to assign all known caps, while `dockerd` itself might not have all caps already.

Background: [Talos](https://www.talos.dev/) in new version 0.13 started dropping two capabilities (kexec + module loading) from all processes but PID 1. Talos itself doesn't use `dockerd`, but if I launch privileged pod on Kubernetes with `docker:20.10-dind` image, I can't run any privileged container inside:

```
/ # docker run -it --rm --privileged alpine
docker: Error response from daemon: OCI runtime create failed: container_linux.go:380: starting container process caused: apply caps: operation not permitted: unknown.
```

The problem starts in https://github.com/moby/moby/blob/306fa44b7ca59282dc8695e6d169c5b25698d0cb/daemon/exec_linux.go#L25

Which essentially uses the list of capabilities built here:
https://github.com/moby/moby/blob/306fa44b7ca59282dc8695e6d169c5b25698d0cb/oci/caps/utils.go#L23-L37

This list contains every capability present on the host, which might be not be true (as some capabilities might have already been dropped).

Containerd OCI does proper thing:

https://github.com/containerd/containerd/blob/d193dc2b8afb1467255cea5326e9807514f94c0f/pkg/cap/cap_linux.go#L123-L136

I'm happy to send a PR, but what is the best way to solve this in the docker codebase?

**Comment by thaJeztah** [comment]
Thanks for reporting; this looks to be related to https://github.com/opencontainers/runtime-spec/pull/1094, which was intended to address these situations. The idea of that is that "higher level" runtimes (containerd, dockerd) would be able to set the set of capabilities, but keeping the OCI runtime the "single source of truth" to handle unknown/unavailable capabilities.

From your description, it looks like the related change in runc (https://github.com/opencontainers/runc/pull/2854) doesn't address this specific case; wondering if it's currently only ignoring "unknown" capabilities, but not "known, but not available" capabilities.

It's been a while since I looked at that, so I would have to dig into what exactly is happening, but if you are interested in looking into that (for runc), let me know.

**Comment by smira** [comment]
The opencontainers/runtime-spec#1094 seems like a move in the right direction, but it wasn't fully implemented? It seems to only ignore unknown caps which is not quite same as the caps which can't be granted.

But yep, exactly what you said.

My understanding is the bug is clearly in `dockerd` - it tries to grant all capabilities, while it can only grant those capabilities which are in the effective set of `dockerd` (it can't grant more that it has). 

I was asking whether using `cap` from containerd would make sense here.
